libimobiledevice: 1.3.0-date=2023-04-30 -> 1.3.0-unstable-2024-05-20,  libimobiledevice-glue: 1.2.0 -> 1.3.0

C0D3-M4513R opened the pull request:
## Description of changes

* libimobiledevice: 1.3.0-date=2023-04-30 -> 1.3.0-unstable-2024-05-20
* libimobiledevice-glue: 1.2.0 -> 1.3.0

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

superherointj commented:
You have to split in different commits. (And follow the contribution guidelines.)

RossComputerGuy reviewed (changes requested):
Please split the two package updates into their own commit and properly format commits and PR title to something like `libimobiledevice: 1.3.0-date=2023-04-30 -> 1.3.0-date=2024-06-19` and `libimobiledevice-glue: 1.2.0 -> 1.3.0`.

superherointj commented:
Is this `date=` a new format for `unstable`?

RossComputerGuy commented:
I'm not sure, this package has been using it since before I became its maintainer. We always could change it or just keep the flow.

superherointj commented:
> I'm not sure, this package has been using it since before I became its maintainer. We always could change it or just keep the flow.

Thanks for clarifying. I think we should follow the guidelines for contribution unless there is a good reason. And then, the good reason should be documented. I think, this may be just the usual case of `unstable` usage.

superherointj commented:
This package is missing the update script. There is a `gitUpdater` abstraction exactly for this case of rolling git.

C0D3-M4513R committed:
libimobiledevice-glue: 1.2.0 -> 1.3.0

Signed-off-by: C0D3 M4513R <28912031+C0D3-M4513R@users.noreply.github.com>

C0D3-M4513R committed:
libimobiledevice: 1.3.0-date=2023-04-30 -> 1.3.0-unstable-2024-06-19

Signed-off-by: C0D3 M4513R <28912031+C0D3-M4513R@users.noreply.github.com>

C0D3-M4513R commented:
I've removed the `libimobiledevice-glue` bump from this pr, added the `unstableGitUpdater`, made sure `nix-shell maintainers/scripts/update.nix --argstr package libimobiledevice` works and am currently (again) building.

Please note, that the previous build still has not completed for me.
If that's what you meant, when you said "And follow the contribution guidelines", then I'm sorry.
`ofborg-eval` is in this case simply way faster.

I take it, that's all you wanted to happen?

superherointj commented:
> I've removed the `libimobiledevice-glue` bump from this pr

A single PR can have multiple commits. The problem was that you added more than a single package to a single commit.
You can keep `libimobiledevice-glue` here as long as it's a different commit.

> added the `unstableGitUpdater`, made sure `nix-shell maintainers/scripts/update.nix --argstr package libimobiledevice` works and am currently (again) building.

Good. That's how it's supposed to be. Always favor the update script over manual updates. Only in cases of failure, that the package or update script have to be tweaked that you should do manual updates.

> Please note, that the previous build still has not completed for me. If that's what you meant, when you said "And follow the contribution guidelines", then I'm sorry. `ofborg-eval` is in this case simply way faster at building c.a. 454 packages from source.

FYI, eval is different than building a package. If you propose a package that you haven't built, and you don't know it is building, you may break it for everyone. (And no one wants that.)

Realize, I also have to wait for builds. That is what it is. But I do that for way more packages/PRs over that you might go over.

> I take it, that's all you wanted to happen?

:-)
Actually no. But I don't want to overload you.
I think these packages should be migrated to `by-name`.
But it is unrequired atm.

superherointj commented on pkgs/development/libraries/libimobiledevice/default.nix:
```diff
@@ -12,29 +12,24 @@
 
   outputs = [ "out" "dev" ];
 
+  passthru.updateScript = unstableGitUpdater { };
```
`passthru` should be added just before `meta`.

There is a thing called `phases order` which is the order phases are executed.

See manual here:
https://nixos.org/manual/nixpkgs/unstable/#sec-stdenv-phases

Whenever adding attributes, consider where that attribute belongs to, and add at that point of derivation. Errors will appear in sequence when doing this. It's the order of execution.

C0D3-M4513R replied:
> Whenever adding attributes, consider where that attribute belongs to, and add at that point of derivation. Errors will appear in sequence when doing this. It's the order of execution.

I'm sorry, but I do not know which attribute belongs to what phase. Is there any documentation for that, other than searching through the entire nixpkgs source code each time?

superherointj replied:
I just told you the answer.

C0D3-M4513R replied:
> `passthru` should be added just before `meta`
> I'd add `output` just before `passhtru`

so the `meta`, `output` and `passthru` are all in a phase after `configure` ?

superherointj replied:
`preAutoreconf`, `configureFlags` should be before buildPhase, meaning before `nativeBuildInputs`.

Also skip a line between each of these attributes. It aids in reading.

superherointj commented:
> when you said "And follow the contribution guidelines"

There is a certain format to commit messages. There exist automation to do auto sort of things to get your PR right.

Follow this format:

> package_name: oldVersion -> newVersion

Note: I **only** want to point out you are missing the `:` in your commit messages.

superherointj commented on pkgs/development/libraries/libimobiledevice/default.nix:
```diff
@@ -60,6 +49,8 @@ stdenv.mkDerivation rec {
 
   configureFlags = [ "--without-cython" ];
 
+  outputs = [ "out" "dev" ];
+  passthru.updateScript = unstableGitUpdater { };
```
```suggestion
  outputs = [ "out" "dev" ];

  passthru.updateScript = unstableGitUpdater { };

```

superherointj commented:
You got phases order for `libimobiledevice` right now.

C0D3-M4513R commented:
As per your comment earlier, doesn't that mean I should move `preAutoreconf` to before `nativeBuildInputs` and `output` to just before `meta` in `libimobiledevice-glue` then?

superherointj commented:
> As per your comment earlier, doesn't that mean I should move `preAutoreconf` to before `nativeBuildInputs` and `output` to just before `meta` in `libimobiledevice-glue` then?

Yes, it will be the same.

C0D3-M4513R commented:
I'm starting a build again then (this time on another machine, which I'm not actively using).
Let's see how many hours the build will take.

superherointj commented:
> I'm starting a build again then (this time on another machine, which I'm not actively using). Let's see how many hours the build will take.

You can use `@ofborg` to build it here. It's fast.

https://github.com/NixOS/ofborg?tab=readme-ov-file#ofborg

superherointj reviewed (changes requested):
`libimobiledevice-glue` update script is pending

My suggestion: (I have tested and it works.)

> passthru.updateScript = nix-update-script { };

Note: You have to add `nix-update-script` argument to package.

Test with:

> nix-shell maintainers/scripts/update.nix --argstr package libimobiledevice-glue

More information:
https://nixos.org/manual/nixpkgs/unstable/#var-passthru-updateScript

superherointj commented:
Now that update script is working for both packages. You no longer have to do manual updates. But just review the bot updates. I suggest you to **optionally** add yourself as a maintainer of package for getting notified when an update is available. Such that, you can approve those upgrades and have package made available to you sooner.

If you add yourself as a maintainer in a different commit, you can `git revert` that commit directly when you no longer want to be maintainer.

C0D3-M4513R commented:
I am currently not intrested in becoming a maintainer for these packages.
(maybe I'll become one for recaf, but that is not packaged yet [I have a pr branch open to recaf itself])

Are we good to start a ofborg build? And is c.a. 450 packages fine for ofborg?

superherointj commented:
> I am currently not intrested in becoming a maintainer for these packages. (maybe I'll become one for recaf, but that is not packaged yet [I have a pr branch open to recaf itself])
> 
> Are we good to start a ofborg build? And is c.a. 450 packages fine for ofborg?

Up to 500 you target master. And OfBorg can handle it.
But over it, target staging. It's more or less. I'll try to build it on my host. And see if I can build it here.

C0D3-M4513R commented:
@ofborg build

superherointj commented:
> @ofborg build

**I think**, you need to specify the packages that you want to be built. If there is downstream packages that you want to be built, you have to specify it too.

C0D3-M4513R commented:
I think I'm just not a trusted user.  ofborg build only takes in a list of parameters to be passed to nix-build iirc.

RossComputerGuy reviewed (commented):
Thank you @superherointj for reviewing. I'm at work right now but I'll take a look this evening when I'm free.

superherointj commented:
> > Are we good to start a ofborg build? And is c.a. 450 packages fine for ofborg?
> 
> Up to 500, target master and OfBorg can handle it. But over it (like thousands), target staging. It's more or less. I'll attempt to build on my host and post results.

My host failed to build it. Indeed, this build is large. I think, you should target staging. Despite not being that many packages numbers wise...

RossComputerGuy commented:
Set as draft, rebase, retarget, and unmark as draft @C0D3-M4513R

superherointj reviewed (changes requested):
Rebase to staging.

First, mark the PR as draft. (Actually, only verify it is marked as draft. I've marked it as draft for you.)

Then:
https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging

C0D3-M4513R commented:
> > > Are we good to start a ofborg build? And is c.a. 450 packages fine for ofborg?
> > 
> > 
> > Up to 500, target master and OfBorg can handle it. But over it (like thousands), target staging. It's more or less. I'll attempt to build on my host and post results.
> 
> My host failed to build it. Indeed, this build is large. I think, you should target staging. Despite not being that many packages numbers wise...
> 
> Mark it as a draft first. Because odds are you'll get this wrong and ping everyone in nixpkgs. (I'm saying this because people often gets this wrong.)

With instructions as clear as that, it was pretty straight forward.
My host personally is still building this (I was at 153/436. I'm starting a build again now).
My plan to use my laptop backfired though, because 16G Ram+ 16G swap is not enough and I got oom'd. pretty quickly)

superherointj commented:
> With instructions as clear as those in the contributing document, it was pretty straight forward. (thanks for linking) My host personally is still building this (I was at 153/436. I'm starting a build again now). My plan to use my laptop backfired though, because 16G Ram+ 16G swap is not enough and I got oom'd. pretty quickly)

The trick to do larger builds is to lower `max-jobs`, such that you don't run out RAM and swap doesn't happen, after swap happens host degrades to a halt, and then the build doesn't happen.

My host is larger than yours and I still failed to build. I think, it is better to leave this for hydra at this point.

Certify to look up packages and logs at Hydra when build is done:

https://hydra.nixos.org/project/nixpkgs

I'll be merging this PR when eval is done.

C0D3-M4513R commented:
Did I rebase to the right branch? Hydra reports the last staging build to be at `2020-10-05`?

superherointj commented:
> Did I rebase to the right branch?

I think so.

> Hydra reports the last staging build to be at `2020-10-05`?

It is used the other branch. I think, we should ask administrators to disable/remove this. It causes a lot of confusion.

C0D3-M4513R commented:
The naming isn't the best, but I'm scared to try to rename the branch using github's ui.

Also I suppose this supersedes #320292 .

superherointj commented:
> The naming isn't the best, but I'm scared to try to rename the branch using github's ui.

I did not suggest you to rename it at this point in time. Only mentioned it for you awareness. (That you pick a better name for next time.)

> Also I suppose this supersedes #320292 .

Yes, will close it.

superherointj commented:
> Did I rebase to the right branch?

![image](https://github.com/NixOS/nixpkgs/assets/5861043/ba38c8f2-82ff-4746-98e3-2d388c1c7c6a)

I think something is weird.
Let's do the following:

I have named `nixpkgs` as `upstream`. If you have it named something else, use that name.

> git fetch upstream staging
> git reset --hard upstream/staging
> git cherry-pick 07c1f4e1e6be98dddf70a77e29becf56b0639612
> git cherry-pick 35b5a13095d8c1ca631afc03df42dbb65dc121bd

If you do this, you'll have a clean staging for sure. Then, you force push it.

superherointj reviewed (approved):
Tested eval on my host. I expect this to be fine.
I'm merging without waiting for CI eval because it takes forever. And I just tested eval (not full build, can't do full build).

trofi commented:
The build regressed on `clang` and `gcc-14`. Proposed downstream fix as https://github.com/NixOS/nixpkgs/pull/321723
